{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Salmon\n",
    "\n",
    "Salmon is a tool for quantifying the expression of transcripts using RNA-seq data. Salmon uses new algorithms (specifically, coupling the concept of quasi-mapping with a two-phase inference procedure) to provide accurate expression estimates very quickly (i.e. wicked-fast) and while using little memory. Salmon performs its inference using an expressive and realistic model of RNA-seq data that takes into account experimental attributes and biases commonly observed in real RNA-seq data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!wget -O /home/ubuntu/hussainather/omicsedu/data/transcript.fa.gz ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/146/045/GCF_000146045.2_R64/GCF_000146045.2_R64_rna.fna.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Then unzip the transcript.fa.gz file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!gunzip /home/ubuntu/hussainather/omicsedu/data/transcript.fa.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Version Info: Could not resolve upgrade information in the alotted time.\n",
      "Check for upgrades manually at https://combine-lab.github.io/salmon\n",
      "[2018-04-18 01:54:10.085] [jLog] [info] building index\n",
      "RapMap Indexer\n",
      "\n",
      "[Step 1 of 4] : counting k-mers\n",
      "Elapsed time: 0.460229s\n",
      "\n",
      "\u001b[33m\u001b[1m[2018-04-18 01:54:10.546] [jointLog] [warning] Removed 45 transcripts that were sequence duplicates of indexed transcripts.\n",
      "\u001b[00m\u001b[33m\u001b[1m[2018-04-18 01:54:10.546] [jointLog] [warning] If you wish to retain duplicate transcripts, please use the `--keepDuplicates` flag\n",
      "\u001b[00mReplaced 0 non-ATCG nucleotides\n",
      "Clipped poly-A tails from 0 transcripts\n",
      "Building rank-select dictionary and saving to disk done\n",
      "Elapsed time: 0.000625499s\n",
      "Writing sequence data to file . . . done\n",
      "Elapsed time: 0.0052855s\n",
      "[info] Building 32-bit suffix array (length of generalized text is 8793748)\n",
      "Building suffix array . . . success\n",
      "saving to disk . . . done\n",
      "Elapsed time: 0.0212169s\n",
      "done\n",
      "Elapsed time: 0.636463s\n",
      "processed 8000000 positions\n",
      "khash had 8208585 keys\n",
      "saving hash to disk . . . done\n",
      "Elapsed time: 0.488635s\n",
      "[2018-04-18 01:54:20.031] [jLog] [info] done building index\n"
     ]
    }
   ],
   "source": [
    "salmon_index = \"/home/ubuntu/hussainather/omicsedu/test/yeast-salmon.idx\"\n",
    "\n",
    "!salmon index -t /home/ubuntu/hussainather/omicsedu/data/transcript.fa -i $salmon_index --type quasi -k 31\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Version Info: Could not resolve upgrade information in the alotted time.\n",
      "Check for upgrades manually at https://combine-lab.github.io/salmon\n",
      "# salmon (alignment-based) v0.9.1\n",
      "# [ program ] => salmon \n",
      "# [ command ] => quant \n",
      "# [ targets ] => { /home/ubuntu/hussainather/omicsedu/test/yeast-salmon.idx }\n",
      "# [ libType ] => { A }\n",
      "# [ output ] => { /home/ubuntu/hussainather/omicsedu/test }\n",
      "# [ alignments ] => { /home/ubuntu/hussainather/omicsedu/test/SRR5494627.sorted.bam }\n",
      "Logs will be written to /home/ubuntu/hussainather/omicsedu/test/logs\n",
      "Library format { type:single end, relative orientation:none, strandedness:unstranded }\n",
      "\u001b[00m[2018-04-18 02:06:31.713] [jointLog] [info] numQuantThreads = 3\n",
      "\u001b[00mparseThreads = 3\n",
      "Checking that provided alignment files have consistent headers . . . done\n",
      "Populating targets from aln = \"/home/ubuntu/hussainather/omicsedu/test/SRR5494627.sorted.bam\", fasta = \"/home/ubuntu/hussainather/omicsedu/test/yeast-salmon.idx\" . . .\u001b[1m\u001b[41m[2018-04-18 02:06:33.072] [jointLog] [critical] Transcript XV appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:33.072] [jointLog] [critical] Transcript XIII appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:33.072] [jointLog] [critical] Transcript XII appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:33.072] [jointLog] [critical] Transcript XI appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:33.072] [jointLog] [critical] Transcript V appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:33.072] [jointLog] [critical] Transcript XVI appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:33.072] [jointLog] [critical] Transcript II appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:33.072] [jointLog] [critical] Transcript IV appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:33.072] [jointLog] [critical] Transcript III appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:33.072] [jointLog] [critical] Transcript VII appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:33.072] [jointLog] [critical] Transcript IX appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:33.072] [jointLog] [critical] Transcript VI appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:33.072] [jointLog] [critical] Transcript I appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:33.072] [jointLog] [critical] Transcript Mito appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:33.072] [jointLog] [critical] Transcript XIV appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:33.072] [jointLog] [critical] Transcript X appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:33.072] [jointLog] [critical] Transcript VIII appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:33.072] [jointLog] [critical] Please provide a reference FASTA file that includes all targets present in the BAM header\n",
      "If you have access to the genome FASTA and GTF used for alignment \n",
      "consider generating a transcriptome fasta using a command like: \n",
      "gffread -w output.fa -g genome.fa genome.gtf\n",
      "you can find the gffread utility at (http://ccb.jhu.edu/software/stringtie/gff.shtml)\n",
      "\u001b[00mVersion Info: Could not resolve upgrade information in the alotted time.\n",
      "Check for upgrades manually at https://combine-lab.github.io/salmon\n",
      "# salmon (alignment-based) v0.9.1\n",
      "# [ program ] => salmon \n",
      "# [ command ] => quant \n",
      "# [ targets ] => { /home/ubuntu/hussainather/omicsedu/test/yeast-salmon.idx }\n",
      "# [ libType ] => { A }\n",
      "# [ output ] => { /home/ubuntu/hussainather/omicsedu/test }\n",
      "# [ alignments ] => { /home/ubuntu/hussainather/omicsedu/test/SRR5494628.sorted.bam }\n",
      "Logs will be written to /home/ubuntu/hussainather/omicsedu/test/logs\n",
      "Library format { type:single end, relative orientation:none, strandedness:unstranded }\n",
      "\u001b[00m[2018-04-18 02:06:33.360] [jointLog] [info] numQuantThreads = 3\n",
      "\u001b[00mparseThreads = 3\n",
      "Checking that provided alignment files have consistent headers . . . done\n",
      "Populating targets from aln = \"/home/ubuntu/hussainather/omicsedu/test/SRR5494628.sorted.bam\", fasta = \"/home/ubuntu/hussainather/omicsedu/test/yeast-salmon.idx\" . . .\u001b[1m\u001b[41m[2018-04-18 02:06:33.902] [jointLog] [critical] Transcript XV appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:33.902] [jointLog] [critical] Transcript XIII appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:33.902] [jointLog] [critical] Transcript XII appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:33.902] [jointLog] [critical] Transcript XI appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:33.902] [jointLog] [critical] Transcript V appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:33.902] [jointLog] [critical] Transcript XVI appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:33.902] [jointLog] [critical] Transcript II appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:33.902] [jointLog] [critical] Transcript IV appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:33.902] [jointLog] [critical] Transcript III appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:33.902] [jointLog] [critical] Transcript VII appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:33.902] [jointLog] [critical] Transcript IX appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:33.902] [jointLog] [critical] Transcript VI appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:33.902] [jointLog] [critical] Transcript I appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:33.902] [jointLog] [critical] Transcript Mito appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:33.902] [jointLog] [critical] Transcript XIV appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:33.902] [jointLog] [critical] Transcript X appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:33.902] [jointLog] [critical] Transcript VIII appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:33.902] [jointLog] [critical] Please provide a reference FASTA file that includes all targets present in the BAM header\n",
      "If you have access to the genome FASTA and GTF used for alignment \n",
      "consider generating a transcriptome fasta using a command like: \n",
      "gffread -w output.fa -g genome.fa genome.gtf\n",
      "you can find the gffread utility at (http://ccb.jhu.edu/software/stringtie/gff.shtml)\n",
      "\u001b[00mVersion Info: Could not resolve upgrade information in the alotted time.\n",
      "Check for upgrades manually at https://combine-lab.github.io/salmon\n",
      "# salmon (alignment-based) v0.9.1\n",
      "# [ program ] => salmon \n",
      "# [ command ] => quant \n",
      "# [ targets ] => { /home/ubuntu/hussainather/omicsedu/test/yeast-salmon.idx }\n",
      "# [ libType ] => { A }\n",
      "# [ output ] => { /home/ubuntu/hussainather/omicsedu/test }\n",
      "# [ alignments ] => { /home/ubuntu/hussainather/omicsedu/test/SRR5494629.sorted.bam }\n",
      "Logs will be written to /home/ubuntu/hussainather/omicsedu/test/logs\n",
      "Library format { type:single end, relative orientation:none, strandedness:unstranded }\n",
      "\u001b[00m[2018-04-18 02:06:34.301] [jointLog] [info] numQuantThreads = 3\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[00mparseThreads = 3\n",
      "Checking that provided alignment files have consistent headers . . . done\n",
      "Populating targets from aln = \"/home/ubuntu/hussainather/omicsedu/test/SRR5494629.sorted.bam\", fasta = \"/home/ubuntu/hussainather/omicsedu/test/yeast-salmon.idx\" . . .\u001b[1m\u001b[41m[2018-04-18 02:06:34.829] [jointLog] [critical] Transcript XV appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:34.829] [jointLog] [critical] Transcript XIII appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:34.829] [jointLog] [critical] Transcript XII appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:34.829] [jointLog] [critical] Transcript XI appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:34.829] [jointLog] [critical] Transcript V appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:34.829] [jointLog] [critical] Transcript XVI appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:34.829] [jointLog] [critical] Transcript II appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:34.829] [jointLog] [critical] Transcript IV appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:34.829] [jointLog] [critical] Transcript III appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:34.829] [jointLog] [critical] Transcript VII appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:34.829] [jointLog] [critical] Transcript IX appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:34.829] [jointLog] [critical] Transcript VI appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:34.829] [jointLog] [critical] Transcript I appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:34.829] [jointLog] [critical] Transcript Mito appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:34.829] [jointLog] [critical] Transcript XIV appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:34.829] [jointLog] [critical] Transcript X appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:34.829] [jointLog] [critical] Transcript VIII appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:34.829] [jointLog] [critical] Please provide a reference FASTA file that includes all targets present in the BAM header\n",
      "If you have access to the genome FASTA and GTF used for alignment \n",
      "consider generating a transcriptome fasta using a command like: \n",
      "gffread -w output.fa -g genome.fa genome.gtf\n",
      "you can find the gffread utility at (http://ccb.jhu.edu/software/stringtie/gff.shtml)\n",
      "\u001b[00mVersion Info: Could not resolve upgrade information in the alotted time.\n",
      "Check for upgrades manually at https://combine-lab.github.io/salmon\n",
      "# salmon (alignment-based) v0.9.1\n",
      "# [ program ] => salmon \n",
      "# [ command ] => quant \n",
      "# [ targets ] => { /home/ubuntu/hussainather/omicsedu/test/yeast-salmon.idx }\n",
      "# [ libType ] => { A }\n",
      "# [ output ] => { /home/ubuntu/hussainather/omicsedu/test }\n",
      "# [ alignments ] => { /home/ubuntu/hussainather/omicsedu/test/SRR5494630.sorted.bam }\n",
      "Logs will be written to /home/ubuntu/hussainather/omicsedu/test/logs\n",
      "Library format { type:single end, relative orientation:none, strandedness:unstranded }\n",
      "\u001b[00m[2018-04-18 02:06:35.239] [jointLog] [info] numQuantThreads = 3\n",
      "\u001b[00mparseThreads = 3\n",
      "Checking that provided alignment files have consistent headers . . . done\n",
      "Populating targets from aln = \"/home/ubuntu/hussainather/omicsedu/test/SRR5494630.sorted.bam\", fasta = \"/home/ubuntu/hussainather/omicsedu/test/yeast-salmon.idx\" . . .\u001b[1m\u001b[41m[2018-04-18 02:06:35.764] [jointLog] [critical] Transcript XV appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:35.765] [jointLog] [critical] Transcript XIII appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:35.765] [jointLog] [critical] Transcript XII appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:35.765] [jointLog] [critical] Transcript XI appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:35.765] [jointLog] [critical] Transcript V appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:35.765] [jointLog] [critical] Transcript XVI appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:35.765] [jointLog] [critical] Transcript II appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:35.765] [jointLog] [critical] Transcript IV appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:35.765] [jointLog] [critical] Transcript III appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:35.765] [jointLog] [critical] Transcript VII appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:35.765] [jointLog] [critical] Transcript IX appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:35.765] [jointLog] [critical] Transcript VI appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:35.765] [jointLog] [critical] Transcript I appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:35.765] [jointLog] [critical] Transcript Mito appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:35.765] [jointLog] [critical] Transcript XIV appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:35.765] [jointLog] [critical] Transcript X appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:35.765] [jointLog] [critical] Transcript VIII appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:35.765] [jointLog] [critical] Please provide a reference FASTA file that includes all targets present in the BAM header\n",
      "If you have access to the genome FASTA and GTF used for alignment \n",
      "consider generating a transcriptome fasta using a command like: \n",
      "gffread -w output.fa -g genome.fa genome.gtf\n",
      "you can find the gffread utility at (http://ccb.jhu.edu/software/stringtie/gff.shtml)\n",
      "\u001b[00mVersion Info: Could not resolve upgrade information in the alotted time.\n",
      "Check for upgrades manually at https://combine-lab.github.io/salmon\n",
      "# salmon (alignment-based) v0.9.1\n",
      "# [ program ] => salmon \n",
      "# [ command ] => quant \n",
      "# [ targets ] => { /home/ubuntu/hussainather/omicsedu/test/yeast-salmon.idx }\n",
      "# [ libType ] => { A }\n",
      "# [ output ] => { /home/ubuntu/hussainather/omicsedu/test }\n",
      "# [ alignments ] => { /home/ubuntu/hussainather/omicsedu/test/SRR5494631.sorted.bam }\n",
      "Logs will be written to /home/ubuntu/hussainather/omicsedu/test/logs\n",
      "Library format { type:single end, relative orientation:none, strandedness:unstranded }\n",
      "\u001b[00m[2018-04-18 02:06:36.179] [jointLog] [info] numQuantThreads = 3\n",
      "\u001b[00mparseThreads = 3\n",
      "Checking that provided alignment files have consistent headers . . . done\n",
      "Populating targets from aln = \"/home/ubuntu/hussainather/omicsedu/test/SRR5494631.sorted.bam\", fasta = \"/home/ubuntu/hussainather/omicsedu/test/yeast-salmon.idx\" . . .\u001b[1m\u001b[41m[2018-04-18 02:06:36.720] [jointLog] [critical] Transcript XV appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:36.720] [jointLog] [critical] Transcript XIII appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:36.720] [jointLog] [critical] Transcript XII appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:36.720] [jointLog] [critical] Transcript XI appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:36.720] [jointLog] [critical] Transcript V appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:36.720] [jointLog] [critical] Transcript XVI appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:36.720] [jointLog] [critical] Transcript II appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:36.720] [jointLog] [critical] Transcript IV appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:36.720] [jointLog] [critical] Transcript III appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:36.720] [jointLog] [critical] Transcript VII appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:36.720] [jointLog] [critical] Transcript IX appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:36.720] [jointLog] [critical] Transcript VI appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:36.720] [jointLog] [critical] Transcript I appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:36.720] [jointLog] [critical] Transcript Mito appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:36.720] [jointLog] [critical] Transcript XIV appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:36.720] [jointLog] [critical] Transcript X appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:36.720] [jointLog] [critical] Transcript VIII appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:36.720] [jointLog] [critical] Please provide a reference FASTA file that includes all targets present in the BAM header\n",
      "If you have access to the genome FASTA and GTF used for alignment \n",
      "consider generating a transcriptome fasta using a command like: \n",
      "gffread -w output.fa -g genome.fa genome.gtf\n",
      "you can find the gffread utility at (http://ccb.jhu.edu/software/stringtie/gff.shtml)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[00mVersion Info: Could not resolve upgrade information in the alotted time.\n",
      "Check for upgrades manually at https://combine-lab.github.io/salmon\n",
      "# salmon (alignment-based) v0.9.1\n",
      "# [ program ] => salmon \n",
      "# [ command ] => quant \n",
      "# [ targets ] => { /home/ubuntu/hussainather/omicsedu/test/yeast-salmon.idx }\n",
      "# [ libType ] => { A }\n",
      "# [ output ] => { /home/ubuntu/hussainather/omicsedu/test }\n",
      "# [ alignments ] => { /home/ubuntu/hussainather/omicsedu/test/SRR5494632.sorted.bam }\n",
      "Logs will be written to /home/ubuntu/hussainather/omicsedu/test/logs\n",
      "Library format { type:single end, relative orientation:none, strandedness:unstranded }\n",
      "\u001b[00m[2018-04-18 02:06:37.118] [jointLog] [info] numQuantThreads = 3\n",
      "\u001b[00mparseThreads = 3\n",
      "Checking that provided alignment files have consistent headers . . . done\n",
      "Populating targets from aln = \"/home/ubuntu/hussainather/omicsedu/test/SRR5494632.sorted.bam\", fasta = \"/home/ubuntu/hussainather/omicsedu/test/yeast-salmon.idx\" . . .\u001b[1m\u001b[41m[2018-04-18 02:06:37.669] [jointLog] [critical] Transcript XV appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:37.669] [jointLog] [critical] Transcript XIII appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:37.669] [jointLog] [critical] Transcript XII appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:37.669] [jointLog] [critical] Transcript XI appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:37.669] [jointLog] [critical] Transcript V appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:37.669] [jointLog] [critical] Transcript XVI appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:37.669] [jointLog] [critical] Transcript II appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:37.669] [jointLog] [critical] Transcript IV appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:37.669] [jointLog] [critical] Transcript III appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:37.669] [jointLog] [critical] Transcript VII appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:37.669] [jointLog] [critical] Transcript IX appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:37.669] [jointLog] [critical] Transcript VI appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:37.669] [jointLog] [critical] Transcript I appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:37.669] [jointLog] [critical] Transcript Mito appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:37.669] [jointLog] [critical] Transcript XIV appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:37.669] [jointLog] [critical] Transcript X appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:37.669] [jointLog] [critical] Transcript VIII appeared in the BAM header, but was not in the provided FASTA file\n",
      "\u001b[00m\u001b[1m\u001b[41m[2018-04-18 02:06:37.669] [jointLog] [critical] Please provide a reference FASTA file that includes all targets present in the BAM header\n",
      "If you have access to the genome FASTA and GTF used for alignment \n",
      "consider generating a transcriptome fasta using a command like: \n",
      "gffread -w output.fa -g genome.fa genome.gtf\n",
      "you can find the gffread utility at (http://ccb.jhu.edu/software/stringtie/gff.shtml)\n",
      "\u001b[00m"
     ]
    }
   ],
   "source": [
    "from pandas import read_csv\n",
    "\n",
    "RNASeqSRARunTableFile='/home/ubuntu/hussainather/omicsedu/data/RNASeqSRA.tsv'\n",
    "RNASeqSRATable = read_csv(RNASeqSRARunTableFile, delimiter='\\t')\n",
    "RNASeqoutrun = (RNASeqSRATable[\"Run\"])\n",
    "RNASeqBam = \"test/\" + RNASeqoutrun + \".sorted.bam\"\n",
    "\n",
    "for index, individual in enumerate(RNASeqBam):\n",
    "    !salmon quant --targets $salmon_index -l A --output /home/ubuntu/hussainather/omicsedu/test -a /home/ubuntu/hussainather/omicsedu/$individual"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
